Time based published pages

tvdeyen opened the pull request:
**Work in progress**

# Time based published pages

Yes, finally!

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://time-based-published-pages.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin' target='_blank'>Open stage</a> or <a href='http://time-based-published-pages.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin/teatro.log' target='_blank'>view logs</a>.

robinboening commented:
This is a great feature and it looks very good to me at first view.

tvdeyen commented:
Apparently while developing this I recognized that this feature is nice on the first glance, but will most likely be very seldom used. 

Unless we have versions for pages, this feature is only useful for not yet published pages. If a page was published once, a user will most likely want to republish changes on a time based manner and this will not be possible now. Additionally this change will make the most common usage of the "normal" public state more complex (two dates instead of one checkbox).

So, I don't think that this is good as it's right now.

tvdeyen commented:
@mamhoff @moritzps I would love to here your feedback on this.

tvdeyen commented:
This is how it looks right now.

![alchemy cms - home 2016-01-18 09-34-22](https://cloud.githubusercontent.com/assets/42868/12386555/ad18bc9e-bdc6-11e5-85ec-e8bc6357fe90.png)

mamhoff commented:
I think the feature is valuable, and the two issues you point out can be remedied: 

1. In order to ease the "easy public flag" usability issue, you could have a checkbox "not public" if  `public_on` is `nil`. Once you activate that checkbox the UI changes to the datepickers, with `public_on` filled out with the current date. 

2. I think the most common use case for the feature is publishing time-based, not editing time-based. I'd say let's wait for a feature request for this. 

3. For pages that should be public forever, it'd be great to have a button "forever" that sets `public_until` to `nil`.

moritzps commented:
Time based editing needs versioning and you are right @tvdeyen that this is a bigger issue we should take care of later. But for now this feature (defining from when to when the page is published) is already 
a great enhancement.

The normal case will be that a user wants to publish / unpublish a page. Setting the dates is more unlikely but it does happen. I think the UI should be:

- If the user hits the `publish`-icon in the toolbar of the normal page edit mode the checkbox should be set to active and the avalable_on date set to `now`. 
- If the user activates the checkbox in the settings panel the datepicker should beneath the checkbox and the available_on date should be filled with now. The until date should not be filled.

See the mockup to get an idea: ![quick introduction to moqups moqups 2016-01-20 11-27-37](https://cloud.githubusercontent.com/assets/1819792/12446251/4786ead2-bf69-11e5-8f36-d59bfe9cba78.png)
Anyhow: Great addition, thanks!

tvdeyen commented:
@mamhoff thanks for the feedback

@moritzps great idea and thanks for the mockup!

tvdeyen commented:
@moritzps @mamhoff this was tricky and I needed to add a Javascript date formatting library. There is no other way to share the same date format between Ruby and Javascript. I'm not happy with this, but the feature now works.

[![Quickcast](https://cloud.githubusercontent.com/assets/42868/12764655/59265cb8-c9fb-11e5-8582-e9635f6f9fb1.png)](http://quick.as/xvhnzyr)

I really don't know if this is worth all of this, but anyhow the feature is now done. I still need to add some test, though....

Feedback welcome!

tvdeyen commented:
this is a video: click it :)

tvdeyen commented:
So

- Added tests 
- Squashed & reworded commits
- Rebased

:beer: :tada: :sunflower: 

Although the hound is wild and code climate shows wrong results, this is ready to be merged. @mamhoff

mamhoff commented on app/assets/stylesheets/alchemy/forms.scss:
```diff
@@ -146,7 +146,7 @@ form {
   }
 
-  .control_group {
+  .control_group, .control-group {
```
Wirklich? Kannst du das nicht im Markup verändern?

tvdeyen replied:
As part of this PR a bit out of scope. But I could just use the underscore version in this PR

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+  padding-top: 2px;
+  position: absolute;
+  z-index: 9999;
+  -moz-box-sizing: border-box;
```
You could actually leave that out. `box-sizing` should be understood by every browser these days, especially firefox.

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+  -khtml-user-select: none;
+  -moz-user-select: none;
+  -ms-user-select: none;
+  -o-user-select: none;
```
These ones can also be done without prefixes: http://caniuse.com/#feat=user-select-none

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+}
+
+.xdsoft_datetimepicker * {
+  -moz-box-sizing: border-box;
```
Hound is right.

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+
+.xdsoft_datetimepicker .xdsoft_datepicker.active+.xdsoft_timepicker {
+  margin-top: 8px;
+  margin-bottom: 3px
```
Hound is right.

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+    font-family: 'Alchemy Icons';
+    font-size: 16px;
+    line-height: 18px;
+    content: '\e00a';
```
Ah well, yeah, whatever, I don't mind.

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+
+  &:before {
+    border-width: 0 8px 8px 8px;
+    border-color: transparent transparent $text-color transparent;
```
OK, Hound, you can get this one. I won't mind.

tvdeyen replied:
Really, @houndci ?

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+
+  &:before {
+    border-width: 8px 8px 0 8px;
+    border-color: $text-color transparent transparent transparent;
```
Wooof.

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+}
+
+.xdsoft_datetimepicker .xdsoft_label:hover i {
+  opacity: 1.0;
```
Woof. Whatever, @tvdeyen, you could correct it, but I don't mind for merging.

mamhoff commented on app/assets/stylesheets/alchemy/jquery.datetimepicker.scss:
```diff
@@ -0,0 +1,545 @@
+  /* IE10+ */
+  background: linear-gradient(to bottom, #fff 0%, #f4f8fa 73%);
+  /* W3C */
+  filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#fff', endColorstr='#f4f8fa',GradientType=0 );
```
Damn. Now here, I fully, wholeheartedly, agree with Hound about the utter uselessness of one space between parens. I'd merge anyways.

tvdeyen commented:
@mamhoff actually this is not my CSS code, it's from the datepicker library. I could change all of this, that's right, especially because it will save some bytes. But this is the thing with vendor stuff, it's not always the way you wanted it to be.

And I will disable hound CSS checks. They are annoying and not helpful.

mamhoff commented on app/helpers/alchemy/admin/base_helper.rb:
```diff
@@ -333,38 +333,41 @@ def new_asset_path_with_session_information(asset_type)
+      def alchemy_datepicker(object, method, html_options = {})
+        type = html_options.delete(:type) || 'date'
+        date = html_options.delete(:value) || object.send(method.to_sym).presence
+        date = Time.parse(date) if date.is_a?(String)
```
Hound is definitely right on this one. Please use `Time.zone.parse`. :ice_cream:

tvdeyen replied:
I would even remove this feature. We always pass in Date objects into this helper, never Strings.

mamhoff commented on app/models/alchemy/page/page_scopes.rb:
```diff
@@ -100,5 +96,32 @@ module Page::PageScopes
+          "#{table_name}.public_on <= ? AND".freeze,
+          "(#{table_name}.public_until IS NULL".freeze,
+          "OR #{table_name}.public_until >= ?)".freeze
+        ].join(' '.freeze), current_time, current_time)
```
Nice! :+1:

tvdeyen replied:
Haha, I don't know if I should laugh or cry.

mamhoff commented on app/views/alchemy/admin/pages/_page_for_links.html.erb:
```diff
@@ -18,9 +18,7 @@
-      <span class="page_status <%= page_for_links.public ? 'public' : 'not_public' %>" title="<%= page_for_links.status_title(:public) %>"></span>
-      <span class="page_status <%= page_for_links.visible ? 'visible' : 'not_visible' %>" title="<%= page_for_links.status_title(:visible) %>"></span>
-      <span class="page_status <%= page_for_links.restricted ? 'restricted' : 'not_restricted' %>" title="<%= page_for_links.status_title(:restricted) %>"></span>
+      <%= render 'alchemy/admin/pages/page_infos', page: page_for_links %>
```
Nice extraction. I'd kinda say it could be it's own PR, but I'll merge.

tvdeyen replied:
Yeah, it would be better it's own PR

mamhoff commented on config/locales/alchemy.de.yml:
```diff
@@ -385,6 +385,7 @@ de:
     file: "Datei"
     file_rename_error: "Datei konnte nicht umbenannt werden."
     files: "Dateien"
+    from: 'von'
```
Should these eventually be namespaced to something like `time_picker.from` or something similar? Again, this is not something I'd block upon.

tvdeyen replied:
You mean this "too generic translation key"? :D Yes, you are right

mamhoff commented on db/migrate/20160108174834_add_timebased_publishing_columns_to_pages.rb:
```diff
@@ -0,0 +1,34 @@
+
+    Alchemy::Page.published.each do |page|
+      page.update_column(public_on: page.published_at)
+      say "Updated #{page.name} public state"
```
Nice!

mamhoff commented on lib/alchemy/test_support/factories/page_factory.rb:
```diff
@@ -21,13 +21,13 @@
       page_layout { language.page_layout }
       language_root true
-      public true
+      public_on { Time.current }
```
Hound, this is not important.

tvdeyen replied:
Yeah, we really need to relax hound. Take the solidus approach

mamhoff commented on spec/controllers/pages_controller_spec.rb:
```diff
@@ -127,18 +140,63 @@ module Alchemy
+        alchemy_get :show, urlname: still_public_page.urlname
+        expect(response).to be_success
+      end
+    end
```
Is there a spec for `requesting an expired page`?

tvdeyen replied:
Yes. L157 above. "Requesting a no longer public page"

mamhoff commented:
Look, there's a couple of nitpicks here, but none of them would block me from merging. You decide whether you want to amend, otherwise: Merge date tomorrow.

tvdeyen commented:
The tests are now failing, because of the extracted #924

tvdeyen commented:
Closing in favor of soon the released [Page Versions feature](https://github.com/AlchemyCMS/alchemy_cms/tree/page-versions).
